Ensure we apply nested closure effects in the right order

trueadm opened the pull request:
Release notes: none

Ensure we apply the nested closure effects in the right order. This was the underlying bug after doing much digging.

Furthermore, I added case where this fails. I also found that the ReactElement code was causing issues at one point (I forgot to add the props, type, ref and key as dependencies), but that wasn't needed in the end for this fix, but I thought it best to keep those fixes in.

I also had to make a change to the JSON tester as it was incorrectly failing a test because the function closures were not 1:1, which they're not meant to be.

cblappert commented on src/react/utils.js:
```diff
@@ -566,6 +566,46 @@ export function evaluateWithNestedEffects(
+  result[0] = sc;
+  result[1] = pg;
+
+  if (pb) {
```
This should never be null. Nor should `pp` or `po`

sebmarkbage commented on src/react/utils.js:
```diff
@@ -566,6 +566,46 @@ export function evaluateWithNestedEffects(
+// and created objects of both effects combined into a single effects
+// also ensure use the value and generator from the target effect are
+// used in the new effects object
+function createMergedEffects(realm: Realm, target: Effects, source: Effects): Effects {
```
It seems pretty leaky to have this in the React utils. I've change the structure of this in the past and probably will again so would be nice to have this centralized... if this is a legit technique.

It should also be named something else because for conflicts it'll prefer one effect over another. Effects can't just be arbitrarily joined since they have conflicts.

cblappert commented on src/react/utils.js:
```diff
@@ -566,6 +566,46 @@ export function evaluateWithNestedEffects(
+  if (pb) {
+    pb.forEach((val, key, m) => rb.set(key, val));
+  }
+  sb.forEach((val, key, m) => rb.set(key, val));
```
The way this is written, if target and source overwrite the same value, we use the value from source not target, please document this.

sebmarkbage replied:
This can override a binding that already exists.

I think we should forbid any of these effects to have binding and property modifications unless they're part of a newly created object (so and po respectively). That way there can't be a conflict.

cblappert commented on src/react/utils.js:
```diff
@@ -566,6 +566,46 @@ export function evaluateWithNestedEffects(
+  return result;
+}
+
+export function mergeNestedEffects(realm: Realm, target: Effects, nestedEffects: Array<Effects>): Effects {
```
Please document here at a high level what you intend to happen in cases where the nested effects modify overlapping sets of properties/bindings with different values.

NTillmann commented on src/serializer/ResidualReactElementSerializer.js:
```diff
@@ -139,53 +144,60 @@ export class ResidualReactElementSerializer {
+          reactElementAstNode = this._serializeReactElementToJSXElement(value, reactElement);
+        } else if (this.reactOutput === "create-element") {
+          let createElement = this._getReactCreateElementValue();
+          originalCreateElementIdentifier = this.residualHeapSerializer.serializeValue(createElement);
```
Technically, `createElement` is also a dependency.

trueadm commented:
I've addressed all feedback (I think) that was in the scope of this PR and also included a more detailed description plus small refactors and comments in the code so there's more clarity as to why I did it this way.

trueadm commented:
I found the correct fix after more digging. I was apply the nest closure effects in the wrong order.

gaearon commented on src/utils/json.js:
```diff
@@ -53,7 +53,7 @@ export function mergeAdacentJSONTextNodes(node: JSON, includeFunctionBodies: boo
       if (typeof value === "function") {
         if (!includeFunctionBodies) {
-          node[key] = value.toString();
+          node[key] = "function";
```
Seems like you could get rid of `includeFunctionBodies` now.
